math/prime/millerrabinprimalitytest.go

```diff
@@ -14,25 +15,48 @@ import (
 	"github.com/TheAlgorithms/Go/math/modular"
 )
 
-// findD accepts a number and returns the
-// odd number d such that num = 2^r * d - 1
-func findRD(num int64) (int64, int64) {
-	r := int64(0)
+// formatNum accepts a number and returns the
+// odd number d such that num = 2^s * d + 1
+func formatNum(num int64) (int64, int64) {
+	s := int64(0)
 	d := num - 1
 	for num%2 == 0 {
 		d /= 2
-		r++
+		s++
 	}
-	return d, r
+	return d, s
 }
 
-// MillerTest This is the intermediate step that repeats within the
-// miller rabin primality test for better probabilitic chances of
-// receiving the correct result.
-func MillerTest(d, num int64) (bool, error) {
-	random := rand.Int63n(num-1) + 2
+// isTrivial checks if num's primality is easy to determine.
+// If it is, it returns true and num's primality. Otherwise
+// it returns false and false.
+func isTrivial(num int64) (trivial bool, prime bool) {
```

Comment:
```suggestion
func isTrivial(num int64) (bool, bool) {
```

This function is very confusing in its current design - why do we care whether this is trivial or not? We should only care about whether this is a prime or not.

Furthermore, you are using named return types - there is no readability benefit for them usually (in fact I think you're function signature is not right for its purpose). The main benefit of using named return types is pre-allocation (a compile time optimisation) but you are neither assigning any values to these or using them in any specific way making this redundant.

Reply:
> This function is very confusing in its current design - why do we care whether this is trivial or not? We should only care about whether this is a prime or not.

`isTrivial` checks if a number's primality can be trivially determined. For example, all multiples of 2, other than 2 itself, are not primes, so we do not have to run the Miller-Rabin test in that case (also you can't write even numbers as `2^s * d + 1`). Also, the primality of 2 and 3 can be determined trivially as prime. If we can determine the primality without running the tests, the function will return true and the primality of the number. Otherwise, we have to run the test since we can't determine it's primality, so it returns false and false in that case.

> Furthermore, you are using named return types - there is no readability benefit for them usually (in fact I think you're function signature is not right for its purpose). The main benefit of using named return types is pre-allocation (a compile time optimisation) but you are neither assigning any values to these or using them in any specific way making this redundant.

Named returned types appear in the docs, so they are useful as documentation.

Reply:
I'd argue that the SRP is not violated here because the purpose of this function is not to determine primality, but only **try** to do that. And it returns the result and the flag if it has succeeded or not. This is a common pattern in other languages for operations that can fail (e.g. parsing a string to a number)

Reply:
Fair enough, if you disagree then majority rules. If you're talking about the pattern of result and a flag, then it still doesn't follow idiomatic Go calling structure (`res, ok` rather than `ok, res`).

Anyways, I feel like this is a redundant thing since majority agrees but I still think SRP is violated. The `ok` flag is valid pattern in Go as well but trivial flag is by its nature testing something else rather than testing whether this is a prime (just thinking of the naming of the function itself points that out) so the pattern doesn't really apply here - note that there is separate condition testing for completely separate thing - which I think violates the principle but... I concede.

Reply:
Changing it to follow the comma ok idiom is a good idea, will make the changes.

Reply:
@tjgurwara99 [Here](https://github.com/golovers/effective-go#named-result-parameters) is an excerpt from a document I quite like, which explains my usage of named returns.

Reply:
Also, I made the code more concise using named returns.

Reply:
> @tjgurwara99 [Here](https://github.com/golovers/effective-go#named-result-parameters) is an excerpt from a document I quite like, which explains my usage of named returns.

Will take a look 😄